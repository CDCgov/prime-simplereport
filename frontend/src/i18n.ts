import i18n from "i18next";
import { initReactI18next } from "react-i18next";

import { en } from "./lang/en";
import { es } from "./lang/es";
import { eo } from "./lang/eo";

export const resources = { en, es, eo };

i18n.use(initReactI18next).init({
  resources,
  lng: localStorage.getItem("language") || "en",
  interpolation: {
    escapeValue: false,
  },
});

export default i18n;
