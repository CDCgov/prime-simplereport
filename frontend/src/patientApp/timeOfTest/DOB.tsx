import React, {
  FormEvent,
  useEffect,
  useState,
  useRef,
  MouseEventHandler,
} from "react";
import { connect, useDispatch, useSelector } from "react-redux";
import { Redirect } from "react-router-dom";
import moment from "moment";
import { useTranslation } from "react-i18next";

import iconSprite from "../../../node_modules/uswds/dist/img/sprite.svg";
import Button from "../../app/commonComponents/Button/Button";
import TextInput from "../../app/commonComponents/TextInput";
import { setPatient, updateOrganization } from "../../app/store";
import { PxpApi } from "../PxpApiService";
import Alert from "../../app/commonComponents/Alert";

const DOB = () => {
  const { t } = useTranslation();

  const dispatch = useDispatch();
  const [birthDate, setBirthDate] = useState("");
  const [birthDateError, setBirthDateError] = useState("");
  const [birthDateHidden, setBirthDateHidden] = useState(true);
  const [linkExpiredError, setLinkExpiredError] = useState(false);
  const dobRef = useRef<HTMLInputElement>(null);
  const plid = useSelector((state: any) => state.plid);
  const patient = useSelector((state: any) => state.patient);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    dobRef?.current?.focus();
  }, []);

  const validPattern = new RegExp("([0-9]{1,2}/[0-9]{1,2}/[0-9]{4})");

  const validateBirthDate = () => {
    const date = moment(birthDate, "MM/DD/YYYY");
    if (!validPattern.test(birthDate)) {
      setBirthDateError(t("testResult.dob.invalidFormat"));
      dobRef?.current?.focus();
      return false;
    } else if (date.year() < 1900 || date.year() > moment().year()) {
      setBirthDateError(t("testResult.dob.invalidYear"));
      dobRef?.current?.focus();
      return false;
    } else if (!date.isValid()) {
      setBirthDateError(t("testResult.dob.invalidDate"));
      dobRef?.current?.focus();
      return false;
    } else {
      setBirthDateError("");
      return true;
    }
  };

  const confirmBirthDate = async () => {
    if (!validateBirthDate()) {
      return;
    }

    const date = moment(birthDate, "MM/DD/YYYY");
    setLoading(true);
    try {
      const response = await PxpApi.validateDateOfBirth(
        plid,
        date.format("YYYY-MM-DD")
      );
      dispatch(
        updateOrganization({
          name: response.organizationName,
        })
      );
      dispatch(setPatient(response));
    } catch (error) {
      if (error?.status === 410) {
        setLinkExpiredError(true);
      } else {
        setBirthDateError(t("testResult.dob.error"));
      }
    } finally {
      setLoading(false);
    }
  };

  if (loading) {
    return (
      <main>
        <div className="grid-container maxw-tablet">
          <p className="margin-top-3">{t("testResult.dob.validating")}</p>
        </div>
      </main>
    );
  }
  if (patient?.orderStatus === "COMPLETED") {
    return (
      <Redirect
        to={{
          pathname: "/test-result",
          search: `?plid=${plid}`,
        }}
      />
    );
  } else if (patient?.firstName) {
    return (
      <Redirect
        to={{
          pathname: "/patient-info-confirm",
          search: `?plid=${plid}`,
        }}
      />
    );
  }

  return (
    <>
      <main>
        <div className="grid-container maxw-tablet">
          {!linkExpiredError ? (
            <>
              <p className="margin-top-3">{t("testResult.dob.enterDOB2")}</p>
              <TextInput
                className="width-mobile"
                label={t("testResult.dob.dateOfBirth")}
                name={"birthDate"}
                type={birthDateHidden ? "password" : "text"}
                autoComplete={"on"}
                value={birthDate}
                ariaDescribedBy={"bdayFormat"}
                hintText={t("testResult.dob.format")}
                onBlur={validateBirthDate}
                errorMessage={birthDateError}
                validationStatus={birthDateError ? "error" : undefined}
                onChange={(evt) => setBirthDate(evt.currentTarget.value)}
                inputRef={dobRef}
              />
              <div className="margin-top-1 margin-bottom-2">
                <button
                  className="usa-button usa-button--unstyled margin-top-0"
                  aria-controls="birthDate"
                  onClick={() => setBirthDateHidden(!birthDateHidden)}
                >
                  {birthDateHidden ? "Show my typing" : "Hide my typing"}
                </button>
              </div>
              <Button
                id="dob-submit-button"
                data-testid="dob-submit-button"
                label={t("testResult.dob.submit")}
                onClick={confirmBirthDate}
              />
            </>
          ) : (
            <>
              <p></p>
              <Alert
                type="error"
                title="Link expired"
                body={t("testResult.dob.linkExpired")}
              />
            </>
          )}
        </div>
      </main>
    </>
  );
};

export default connect()(DOB);
