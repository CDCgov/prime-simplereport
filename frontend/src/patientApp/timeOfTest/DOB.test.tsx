import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import "../../i18n";

import { act } from "react-dom/test-utils";

import DOB from "./DOB";

const mockStore = configureStore([]);
const mockContainer = (store: any) => (
  <Provider store={store}>
    <DOB />
  </Provider>
);

jest.mock("../PxpApiService", () => ({
  PxpApi: {
    validateDateOfBirth: (patientLinkId: string, dateOfBirth: string) => {
      return new Promise((res, rej) => {
        if (dateOfBirth === "1987-08-21") {
          res({ organizationName: "Foo" });
        } else {
          rej();
        }
      });
    },
  },
}));

describe("DOB", () => {
  describe("snapshot", () => {
    let component: renderer.ReactTestRenderer;
    beforeEach(() => {
      const store = mockStore({
        plid: "foo",
      });
      component = renderer.create(mockContainer(store));
    });
    it("matches", () => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe("behavior", () => {
    beforeEach(() => {
      const store = mockStore({
        plid: "foo",
      });
      render(mockContainer(store));
    });

    it("Checks for the correct format - dashes aren't accepted", async () => {
      userEvent.type(
        await screen.findByLabelText("Date of birth"),
        "07-12-1987"
      );
      userEvent.click(await screen.findByText("Continue"));
      const error = await screen.findByRole("alert");

      expect(error.textContent).toEqual(
        "Error: The birth date provided is in an incorrect format"
      );
    });

    it("Checks for the correct format - missing slashes aren't accepted", async () => {
      userEvent.type(await screen.findByLabelText("Date of birth"), "07121987");
      userEvent.click(await screen.findByText("Continue"));
      const error = await screen.findByRole("alert");

      expect(error.textContent).toEqual(
        "Error: The birth date provided is in an incorrect format"
      );
    });

    it("Checks to make sure it is actually a possible date", async () => {
      userEvent.type(
        await screen.findByLabelText("Date of birth"),
        "99/99/9999"
      );
      userEvent.click(await screen.findByText("Continue"));
      const error = await screen.findByRole("alert");

      expect(error.textContent).toEqual(
        "Error: The birth date provided is not a possible date"
      );
    });

    it("Checks to make sure it is a date after 1900", async () => {
      userEvent.type(
        await screen.findByLabelText("Date of birth"),
        "08/21/1899"
      );
      userEvent.click(await screen.findByText("Continue"));
      const error = await screen.findByRole("alert");

      expect(error.textContent).toEqual(
        "Error: Birth date must be after 1900 and before the current year"
      );
    });

    it("Checks to make sure it is a date before this year", async () => {
      userEvent.type(
        await screen.findByLabelText("Date of birth"),
        "08/21/2237"
      );
      userEvent.click(await screen.findByText("Continue"));
      const error = await screen.findByRole("alert");

      expect(error.textContent).toEqual(
        "Error: Birth date must be after 1900 and before the current year"
      );
    });

    it("Accepts an otherwise valid date", async () => {
      userEvent.type(
        await screen.findByLabelText("Date of birth"),
        "08/21/1987"
      );
      await act(async () => {
        userEvent.click(await screen.findByText("Continue"));
      });
      expect(screen.queryByRole("alert")).not.toBeInTheDocument();
    });
  });
});
