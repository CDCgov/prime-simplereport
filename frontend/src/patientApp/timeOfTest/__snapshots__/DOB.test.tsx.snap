// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOB snapshot matches 1`] = `
<main>
  <div
    className="grid-container maxw-tablet"
  >
    <p
      className="margin-top-3"
    >
      Enter your date of birth to access your COVID-19 Testing Portal.
    </p>
    <div
      className="usa-form-group width-mobile"
    >
      <label
        aria-describedby="bdayFormat"
        className="usa-label"
        htmlFor="1"
      >
        Date of birth
      </label>
      <span
        className="usa-hint"
      >
        MM/DD/YYYY
      </span>
      <input
        aria-required="false"
        autoComplete="on"
        className="usa-input"
        id="1"
        name="birthDate"
        onBlur={[Function]}
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <div
      className="margin-top-1 margin-bottom-2"
    >
      <button
        aria-controls="birthDate"
        className="usa-button usa-button--unstyled margin-top-0"
        onClick={[Function]}
      >
        Show my typing
      </button>
    </div>
    <button
      className="usa-button"
      id="dob-submit-button"
      onClick={[Function]}
      type="button"
    >
      Continue
    </button>
  </div>
</main>
`;
